File: KCD/HMDestroyItem2Command.m
//
//  HMDestroyItem2Command.m
//  KCD
//
//  Created by Hori,Masaki on 2014/04/29.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMDestroyItem2Command.h"

#import "HMServerDataStore.h"
#import "HMKCSlotItemObject+Extensions.h"
#import "HMKCMaterial.h"


@implementation HMDestroyItem2Command
+ (void)load
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		[HMJSONCommand registerClass:self];
	});
}

+ (BOOL)canExcuteAPI:(NSString *)api
{
	if([api isEqualToString:@"/kcsapi/api_req_kousyou/destroyitem2"]) return YES;
	
	return NO;
}
- (void)execute
{
	HMServerDataStore *store = [HMServerDataStore oneTimeEditor];
	
	NSString *itemsString = self.arguments[@"api_slotitem_ids"];
	NSArray *itemsStrings = [itemsString componentsSeparatedByString:@","];
	NSMutableArray *items = [NSMutableArray new];
	for(id item in itemsStrings) {
		[items addObject:@([item integerValue])];
	}
	
	NSError *error = nil;
	NSArray<HMKCSlotItemObject *> *slotItems = [store objectsWithEntityName:@"SlotItem"
																	  error:&error
															predicateFormat:@"id IN %@", items];
	if(slotItems.count == 0) {
		NSLog(@"SlotItem is invalid.");
		return;
	}
	
	for(id obj in slotItems) {
		[store deleteObject:obj];
	}
	
	//
	error = nil;
	NSArray<HMKCMaterial *> *materials = [store objectsWithEntityName:@"Material" predicate:nil error:&error];
	if(error) {
		[self log:@"Fetch error: %@", error];
		return;
	}
	if(materials.count == 0) {
		NSLog(@"SlotItem is invalid.");
		return;
	}
	
	NSArray *keys = @[@"fuel", @"bull", @"steel", @"bauxite", @"kousokukenzo", @"kousokushuhuku", @"kaihatusizai", @"screw"];
	
	NSArray *gotMaterials = [self.json valueForKeyPath:@"api_data.api_get_material"];
	for(NSInteger i = 0; i < 4; i++) {
		NSInteger current = [[materials[0] valueForKey:keys[i]] integerValue];
		NSInteger increase = [gotMaterials[i] integerValue];
		[materials[0] setValue:@(current + increase) forKey:keys[i]];
	}
}
@end

File: KCD/HMDestroyItem2Command.swift
//
//  HMDestroyItem2Command.swift
//  KCD
//
//  Created by Hori,Masaki on 2015/01/21.
//  Copyright (c) 2015年 Hori,Masaki. All rights reserved.
//

import Cocoa


extension String {
	var integerValue: Int { return toInt() ?? 0 }
}

class HMDestroyItem2Command: HMJSONCommand {
	required override init() {}
	
	override class func load() {
		HMJSONCommand.registerInstance(self())
	}
	override class func canExcuteAPI(api: String) -> Bool {
		return api == "/kcsapi/api_req_kousyou/destroyitem2"
	}
	
	override func execute() {
		let store = HMServerDataStore.oneTimeEditor()
		if let itemsString = arguments["api_slotitem_ids"] as? NSString {
			if let itemsStrings = itemsString.componentsSeparatedByString(",") as? [String] {
				var items: [AnyObject] = itemsStrings.map() {
					(item: String) in
					if let ID = item.toInt() {
						return NSNumber(long: ID)
					} else {
						return -1
					}
				}
				var error: NSError? = nil
				let predicate = NSPredicate(format: "id IN %@", argumentArray:[items])
				let array = store.objectsWithEntityName("SlotItem",
					sortDescriptors: nil,
					predicate: predicate,
					error: &error)
				if array.count == 0 {
					NSLog("SlotItem is invalid")
					if error != nil {
						NSLog("Error: \(error!)")
					}
					return
				}
				for i in array {
					store.managedObjectContext!.deleteObject(i as NSManagedObject)
				}
				
				error = nil
				let materials = store.objectsWithEntityName("Material",
					sortDescriptors: nil,
					predicate: nil,
					error: &error)
				if array.count == 0 {
					NSLog("SlotItem is invalid")
					if error != nil {
						NSLog("Error: \(error!)")
					}
					return
				}
				
				let material = materials[0] as NSManagedObject
				let keys = ["fuel", "bull", "steel", "bauxite", "kousokukenzo", "kousokushuhuku", "kaihatusizai", "screw"]
				if let growsMaterials = json.valueForKeyPath("api_data.api_get_material") as? [AnyObject] {
					for i in 0..<4 {
						let current = material.valueForKey(keys[i])?.integerValue
						let grows = growsMaterials[i].integerValue
						material.setValue(NSNumber(long: current! + grows), forKey:keys[i])
					}
				}
			}
		}
	}
}
